### DISCUSSION TITLE
Can't assign direct access permissions

### DISCUSSION THREAD

**Discussion by pilipeikoko** [pnp/pnpcore#583] [Q&A] [answered]
I want to change direct access for a folder.
Is it possible to create direct permissions like in logic apps?
![image](https://user-images.githubusercontent.com/56627083/137465153-8839d4cd-935c-4c5e-81dc-3e78351ca69f.png)
![image](https://user-images.githubusercontent.com/56627083/137464610-9896de12-f30a-4990-91f2-e25c6358c6d0.png)

All the examples I found are using Microsoft SP ClientContext

**Comment by jansenbe** [comment, marked as answer]
Hi @pilipeikoko ,

This is possible using PnP Core SDK. I assume you've things setup and have been able to get a PnPContext instance (see https://pnp.github.io/pnpcore/using-the-sdk/readme.html). Assume your PnPContext instance variable is named `context` then below code takes a folder, breaks permission inheritance and grants the current user "Full Control" on the folder:

```csharp
var folderToPermission = await context.Web.GetFolderByServerRelativeUrlAsync("/sites/mysite/library/folder1/folder2", x => x.ListItemAllFields);

// Break permission inheritance, get the user and role to assign                
var currentUser = await context.Web.GetCurrentUserAsync();
var roleDefinition = await context.Web.RoleDefinitions.FirstOrDefaultAsync(d => d.Name == "Full Control");

// Batching approach to break role inheritance and apply custom roles for given user, only a single roundtrip needed
await folderToPermission.ListItemAllFields.BreakRoleInheritanceBatchAsync(false, true);
await folderToPermission.ListItemAllFields.AddRoleDefinitionBatchAsync(currentUser.Id, roleDefinition);
await context.ExecuteAsync();
```

**Reply by jansenbe** [reply]
@pilipeikoko : To further clarify things I've done some consistency work in the code base + created the documentation for these options. See https://pnp.github.io/pnpcore/using-the-sdk/security-intro.html for the docs.

**Reply by pilipeikoko** [reply]
Thanks man, that's exactly what I needed.

**Reply by claudiobottari** [reply]
> Hi @pilipeikoko ,
> 
> This is possible using PnP Core SDK. I assume you've things setup and have been able to get a PnPContext instance (see https://pnp.github.io/pnpcore/using-the-sdk/readme.html). Assume your PnPContext instance variable is named `context` then below code takes a folder, breaks permission inheritance and grants the current user "Full Control" on the folder:
> 
> ```cs
> var folderToPermission = await context.Web.GetFolderByServerRelativeUrlAsync("/sites/mysite/library/folder1/folder2", x => x.ListItemAllFields);
> 
> // Break permission inheritance, get the user and role to assign                
> var currentUser = await context.Web.GetCurrentUserAsync();
> var roleDefinition = await context.Web.RoleDefinitions.FirstOrDefaultAsync(d => d.Name == "Full Control");
> 
> // Batching approach to break role inheritance and apply custom roles for given user, only a single roundtrip needed
> await folderToPermission.ListItemAllFields.BreakRoleInheritanceBatchAsync(false, true);
> await folderToPermission.ListItemAllFields.AddRoleDefinitionBatchAsync(currentUser.Id, roleDefinition);
> await context.ExecuteAsync();
> ```

Thank works perfectly for me, I'm able to set permissions for users and sharepoint groups but I'm no t able the retrieve the correct Is when I'm trying to add a role definition for a group obtained through Graph. In that case I only get an Id (string) which is useless for this purpose.... do you have any work-around for this purpose?
